diff --git a/src/kboot.c b/src/kboot.c
index a5bd4f1..1089666 100644
--- a/src/kboot.c
+++ b/src/kboot.c
@@ -1249,17 +1249,23 @@ static int dt_set_acio_tunables(void)
 
         memset(fdt_alias, 0, sizeof(fdt_alias));
         snprintf(fdt_alias, sizeof(fdt_alias), "usb4_%d_rc", i);
+        if (!fdt_get_alias(dt, fdt_alias))
+            snprintf(fdt_alias, sizeof(fdt_alias), "usb4-%d-acio", i);
         dt_copy_acio_tunables(adt_path, fdt_alias, usb4_rc_tunables,
                               sizeof(usb4_rc_tunables) / sizeof(*usb4_rc_tunables));
 
         memset(fdt_alias, 0, sizeof(fdt_alias));
         snprintf(fdt_alias, sizeof(fdt_alias), "usb4_%d_pcie_adapter", i);
+        if (!fdt_get_alias(dt, fdt_alias))
+            snprintf(fdt_alias, sizeof(fdt_alias), "usb4-%d-pcie-adapter", i);
         dt_copy_acio_tunables(adt_path, fdt_alias, usb4_pcie_adapter_tunables,
                               sizeof(usb4_pcie_adapter_tunables) /
                                   sizeof(*usb4_pcie_adapter_tunables));
 
         memset(fdt_alias, 0, sizeof(fdt_alias));
         snprintf(fdt_alias, sizeof(fdt_alias), "usb4_%d_nhi", i);
+        if (!fdt_get_alias(dt, fdt_alias))
+            snprintf(fdt_alias, sizeof(fdt_alias), "usb4-%d-nhi", i);
         dt_copy_acio_tunables(adt_path, fdt_alias, usb4_nhi_tunables,
                               sizeof(usb4_nhi_tunables) / sizeof(*usb4_nhi_tunables));
         dt_copy_usb4_drom(adt_path, fdt_alias);
