diff --git a/src/cpp/src/whisper/word_level_timestamps.cpp b/src/cpp/src/whisper/word_level_timestamps.cpp
index 7b3da124..d1f9212f 100644
--- a/src/cpp/src/whisper/word_level_timestamps.cpp
+++ b/src/cpp/src/whisper/word_level_timestamps.cpp
@@ -327 +327 @@ std::pair<std::vector<std::string>, std::vector<std::vector<int64_t>>> split_toke
-    const std::string replacement_char = u8"\uFFFD";
+    const std::string replacement_char = "\xEF\xBF\xBD";
